services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerEndpoint.java

```diff
@@ -482,12 +470,35 @@ private List<ProtocolMapperModel> getProtocolMappers(List<OID4VCClient> oid4VCCl
                 .toList();
     }
 
-    private String generateNonce() {
-        return SecretGenerator.getInstance().randomString();
+    private String generateCodeForSession(int expiration, AuthenticatedClientSessionModel clientSession) {
+        String codeId = SecretGenerator.getInstance().randomString();
+        String nonce = SecretGenerator.getInstance().randomString();
+        OAuth2Code oAuth2Code = new OAuth2Code(codeId, expiration, nonce, CREDENTIAL_OFFER_URI_CODE_SCOPE, null, null, null,
+                clientSession.getUserSession().getId());
+
+        return OAuth2CodeParser.persistCode(session, clientSession, oAuth2Code);
     }
 
-    private String generateAuthorizationCodeForClientSession(AuthenticatedClientSessionModel clientSessionModel) {
-        int expiration = timeProvider.currentTimeSeconds() + preAuthorizedCodeLifeSpan;
+    private CredentialsOffer getOfferFromSessionCode(String sessionCode) {
+        EventBuilder eventBuilder = new EventBuilder(session.getContext().getRealm(), session,
+                session.getContext().getConnection());
+        OAuth2CodeParser.ParseResult result = OAuth2CodeParser.parseCode(session, sessionCode,
+                session.getContext().getRealm(),
+                eventBuilder);
+        if (result.isExpiredCode() || result.isIllegalCode() || !result.getCodeData().getScope().equals(CREDENTIAL_OFFER_URI_CODE_SCOPE)) {
+            throw new BadRequestException(getErrorResponse(ErrorType.INVALID_TOKEN));
```

Comment:
There is no type ```INVALID_CREDENTIAL```, just  ```INVALID_CREDENTIAL_REQUEST```. Since its not a credential request(yet), and the issue is in the presented code, I thought ```INVALID_TOKEN``` is the best match.

Reply:
Thank you for the clarification.